File: 백준/Gold/2637. 장난감 조립/장난감 조립.js
const readline=require('readline');
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const N=Number(input[iter++])
  const M=Number(input[iter++])

  const link=[...new Array(N+1)].map(item=>[])
  const basic=[...new Array(N+1)].map(item=>0)
  const items={}
  for(let i=0;i<M;i++){
    const [X,Y,K]=input[iter++].split(' ').map(item=>Number(item))
    if (X===N) items[Y]=K
    link[Y].push([X,K])
    basic[X]+=1
  }
  const first=[...basic]

  const leaf=[]
  for(let i=1;i<=N;i++){
    if (basic[i]===0) leaf.push(i)
  }

  const save=[...new Array(N+1)].map(item=>[...new Array(N+1)].map(item=>0))
  while(leaf.length>0){
    const item=leaf.shift()
    const keys=link[item]
    for(let i=0;i<keys.length;i++){
      const [node,cnt]=keys[i]
      basic[node]-=1
      if (basic[node]===0) leaf.push(node)
      if (first[item]===0) {
        save[node][item]+=cnt
      }
      else {
        for(let j=1;j<=N;j++){
          if (save[item][j]!==0) {
            save[node][j]+=save[item][j]*cnt
          }
        }
      }
    }
  }

  const print=[]
  for(let i=1;i<=N;i++){
    if (first[i]===0) print.push(`${i} ${save[N][i]}`)
  }
  console.log(print.join('\n'))
  process.exit()
})

File: 백준/Gold/2637. 장난감 조립/장난감 조립.py
from collections import defaultdict, deque
import sys
input = sys.stdin.readline

N = int(input())
M = int(input())
topologic_parts = [0]*N
needed_parts = [defaultdict(int) for _ in range(N)]
basic_parts = [defaultdict(int) for _ in range(N)]
q = deque()

for _ in range(M) :
  x, y, k = map(int, input().split())
  topologic_parts[x-1] += 1
  needed_parts[y-1][x-1] += k

for i in range(N) :
  if topologic_parts[i] == 0 :
    q.append(i)
    basic_parts[i][i] = 1

while q :
  node = q.popleft()
  for nxt, mul in needed_parts[node].items() :
    for key, val in basic_parts[node].items() :
      basic_parts[nxt][key] += val*mul
    topologic_parts[nxt] -= 1
    if topologic_parts[nxt] == 0 :
      q.append(nxt)

for key, val in sorted(basic_parts[-1].items()) :
  print(key+1, val)
